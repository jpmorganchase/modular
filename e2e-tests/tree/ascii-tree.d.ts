declare module 'ascii-tree' {
  export function generate(input: string): string;
}
