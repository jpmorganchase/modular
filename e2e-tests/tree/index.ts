import fs from 'fs';
import path from 'path';
import asciiTree from 'ascii-tree';
import hash from '@emotion/hash';

function times(str: string, length: number) {
  return Array.from({ length }, () => str).join('');
}

const defaultIgnores = ['node_modules', '.git', '.DS_Store'];
const defaultHashIgnores = [
  // adding lockfiles here because it can be different
  // on different runs; since we install the latest versions
  // of some packages when making a repository
  'yarn.lock',
  'package-lock.json',
  // adding package.json files since versions within these get bumped
  'package.json',
];

function tree(
  _path: string,
  level = 1,
  options = {
    ignores: defaultIgnores,
    hashIgnores: defaultHashIgnores,
  },
): string {
  const stat = fs.statSync(_path);
  if (stat.isDirectory()) {
    const children = fs.readdirSync(_path);
    const dirArr = _path.split('/');
    const dir = dirArr[dirArr.length - 1];
    // todo - should these be sorted?
    // todo - handle symlinks, etc
    return `${times('#', level)}${dir}\n${children
      .filter((child: string) => !options.ignores.includes(child))
      .map((child: string) => tree(path.join(_path, child), level + 1, options))
      .join('\n')}`;
  } else {
    return [
      `${times('#', level)}${path.basename(_path)}`,
      options.hashIgnores.includes(path.basename(_path))
        ? undefined
        : `#${hash(fs.readFileSync(_path, 'utf8'))}`,
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export default function generate(_path: string): string {
  return asciiTree.generate(tree(_path));
}
