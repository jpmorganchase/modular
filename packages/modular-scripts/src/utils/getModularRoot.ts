import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
