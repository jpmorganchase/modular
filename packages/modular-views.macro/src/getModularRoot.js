// keep this in sync with modular-scripts/src/getModularRoot
const path = require('path');
const fs = require('fs-extra');
const findUp = require('find-up');

function isModularRoot(packageJson) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

module.exports = function getModularRoot() {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
};
