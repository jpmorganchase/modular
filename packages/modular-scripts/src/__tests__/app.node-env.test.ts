import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';
import prettier from 'prettier';
import {
  createModularTestContext,
  mockInstallTemplate,
  runModularForTests,
} from '../test/utils';
import getModularRoot from '../utils/getModularRoot';
import rimraf from 'rimraf';
import execa from 'execa';

// These tests must be executed sequentially with `--runInBand`.

// Temporary test context paths set by createModularTestContext()
let tempModularRepo: string;

const modularRoot = getModularRoot();

const packagesPath = path.join(getModularRoot(), 'packages');

function cleanup() {
  rimraf.sync(path.join(packagesPath, 'node-env-app'));
  rimraf.sync(path.join(modularRoot, 'dist/node-env-app'));

  // run yarn so yarn.lock gets reset
  return execa.sync('yarnpkg', ['--silent'], {
    cwd: modularRoot,
  });
}

describe('when working with a NODE_ENV app', () => {
  describe('WHEN building with webpack', () => {
    beforeAll(async () => {
      cleanup();
      runModularForTests(modularRoot, 'add node-env-app --unstable-type app');

      await fs.writeFile(
        path.join(modularRoot, 'packages', 'node-env-app', 'src', 'index.ts'),
        `
      console.log(process.env.NODE_ENV);
      export {};
    `,
      );
      rimraf.sync(path.join(modularRoot, 'dist/node-env-app'));

      runModularForTests(modularRoot, 'build node-env-app');
    });
    afterAll(cleanup);
    it('can build a app', () => {
      expect(tree(path.join(modularRoot, 'dist', 'node-env-app')))
        .toMatchInlineSnapshot(`
        "node-env-app
        ├─ asset-manifest.json #5npfrr
        ├─ favicon.ico #6pu3rg
        ├─ index.html #9j6678
        ├─ logo192.png #1nez7vk
        ├─ logo512.png #1hwqvcc
        ├─ manifest.json #19gah8o
        ├─ package.json
        ├─ robots.txt #1sjb8b3
        └─ static
           └─ js
              ├─ main.a482480b.js #1xwb1v
              ├─ main.a482480b.js.map #4bcy8y
              ├─ runtime-main.97707f9d.js #15lezt9
              └─ runtime-main.97707f9d.js.map #1qz5n9i"
      `);
    });
    it('can generate a js/main.5d879077.js', async () => {
      expect(
        prettier.format(
          String(
            await fs.readFile(
              path.join(
                modularRoot,
                'dist',
                'node-env-app',
                'static',
                'js',
                'main.6f2657b7.js',
              ),
            ),
          ),
          {
            filepath: 'main.6f2657b7.js',
          },
        ),
      ).toMatchSnapshot();
    });
  });
  describe('WHEN building with esbuild', () => {
    beforeAll(async () => {
      await setupNodeEnvApp();
      runModularForTests(tempModularRepo, 'build node-env-app', {
        env: {
          USE_MODULAR_ESBUILD: 'true',
        },
      });
    });

    it('can build a app', () => {
      expect(tree(path.join(tempModularRepo, 'dist', 'node-env-app')))
        .toMatchInlineSnapshot(`
        "node-env-app
        ├─ favicon.ico #6pu3rg
        ├─ index.html #177xfr4
        ├─ logo192.png #1nez7vk
        ├─ logo512.png #1hwqvcc
        ├─ manifest.json #19gah8o
        ├─ package.json
        ├─ robots.txt #1sjb8b3
        └─ static
           └─ js
              ├─ index-L2US6VEL.js #7mn428
              └─ index-L2US6VEL.js.map #sr6neg"
      `);
    });

    it('can generate a js/index-L2US6VEL.js', async () => {
      expect(
        prettier.format(
          String(
            await fs.readFile(
              path.join(
                tempModularRepo,
                'dist',
                'node-env-app',
                'static',
                'js',
                'index-L2US6VEL.js',
              ),
            ),
          ),
          {
            filepath: 'index-L2US6VEL.js',
          },
        ),
      ).toMatchSnapshot();
    });
  });
});

/**
 * Create temp modular repo and add an app called node-env-app
 */
async function setupNodeEnvApp(): Promise<void> {
  tempModularRepo = createModularTestContext();
  mockInstallTemplate(
    path.join(getModularRoot(), '__fixtures__/templates'),
    tempModularRepo,
  );
  runModularForTests(
    tempModularRepo,
    'add node-env-app --unstable-type modular-template-app',
  );

  await fs.writeFile(
    path.join(tempModularRepo, 'packages', 'node-env-app', 'src', 'index.ts'),
    `
      console.log(process.env.NODE_ENV);
      export {};
      `,
  );
}
