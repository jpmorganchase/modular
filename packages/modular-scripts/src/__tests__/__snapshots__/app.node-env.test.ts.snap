// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when working with a NODE_ENV app WHEN building with esbuild can generate a js/index-L2US6VEL.js 1`] = `
"console.log("production");
//# sourceMappingURL=/static/js/index-L2US6VEL.js.map
"
`;

exports[`when working with a NODE_ENV app WHEN building with webpack can generate a js/main.5d879077.js 1`] = `
""use strict";
(self.webpackChunknode_env_app = self.webpackChunknode_env_app || []).push([
  [179],
  {
    560: () => {
      console.log("production");
    },
  },
  (e) => {
    var n;
    (n = 560), e((e.s = n));
  },
]);
//# sourceMappingURL=main.6f2657b7.js.map
"
`;
